packages/next-app-router/src/index.ts

```diff
@@ -0,0 +1,75 @@
+'use client';
+import { createUseFunnel } from '@use-funnel/core';
+import { useSearchParams } from 'next/navigation';
+import { useLayoutEffect, useMemo, useState } from 'react';
+
+export const useFunnel = createUseFunnel(({ id, initialState }) => {
+  const searchParams = useSearchParams();
+  const [state, setState] = useState<Record<string, any>>({});
+  useLayoutEffect(() => {
+    if (typeof window !== 'undefined') {
+      setState(window.history.state);
+    }
+
+    function handlePopState(event: PopStateEvent) {
+      setState(event.state);
+    }
+    window.addEventListener('popstate', handlePopState);
+    return () => {
+      window.removeEventListener('popstate', handlePopState);
+    };
+  }, []);
+
+  const currentStep = searchParams.get(`${id}.step`);
+  const currentContext = state?.[`${id}.context`];
+
+  const currentState = useMemo(() => {
+    return currentStep != null && currentContext != null
+      ? ({
+          step: currentStep,
+          context: currentContext,
+        } as typeof initialState)
+      : initialState;
+  }, [currentStep, currentContext, initialState]);
+
+  const history: (typeof initialState)[] = useMemo(
+    () => state?.[`${id}.histories`] ?? [currentState],
+    [state, currentState],
+  );
+
+  const currentIndex = history.length - 1;
+  return useMemo(
+    () => ({
+      history,
+      currentIndex,
+      currentState,
+      push(newState) {
+        const newSearchParams = new URLSearchParams(searchParams);
+        const newHistoryState = {
+          ...state,
+          [`${id}.context`]: newState.context,
+          [`${id}.histories`]: [...(history ?? []), newState],
+        };
+
+        newSearchParams.set(`${id}.step`, newState.step);
+        window.history.pushState(newHistoryState, '', `?${newSearchParams.toString()}`);
+        setState(newHistoryState);
```

Comment:
```suggestion
        newSearchParams.set(`${id}.step`, newState.step);
        window.history.pushState(newHistoryState, '', `?${newSearchParams.toString()}`);
        setState(prevHistoryState => ({
          ...prevHistoryState,
          [`${id}.context`]: newState.context,
          [`${id}.histories`]: [...(history ?? []), newState],
        }));
```

Instead of removing the dependency on `history.state` in useMemo, I think we could use something like this.

Reply:
What is the "dangerous to cause side effects" inside the setState callback? I didn't understand that.

Reply:
React recommends that setState internal behavior be pure.

Also, for this purpose, react strict mode calls the callback inside setState twice.

This will not affect the production environment, but may cause confusion for developers in the development environment.

In an environment using strictmode, the history stack is stacked twice for a single history.push or history.replace call.

Reply:
Oh, i see. I understanded!